feat: add `import bulk/resume` commands

cristiand391 opened the pull request:
### What does this PR do?
Adds 2 new commands:

## `data import bulk`
take a CSV with fields of an object and bulk insert them into the org:

#### happy path
https://github.com/user-attachments/assets/aaf214c5-c36b-46e3-aee9-44cbebc4edeb


#### failed to import all records

https://github.com/user-attachments/assets/a8699832-e161-4873-bddc-cc6fd9456bd5

#### job failure
![Screenshot 2024-10-16 at 11 08 37 AM](https://github.com/user-attachments/assets/0e70777f-ab1c-47f1-9dea-ed64c32c8aac)

#### job aborted
https://github.com/user-attachments/assets/a099d732-6a40-41f3-830e-76af786dfe11

## `data import resume`
resume an async/timed out import

https://github.com/user-attachments/assets/b05e4849-abb9-4d15-9ee0-983444132f35

### testing instructions

checkout PR locally and `sf plugins link` it, you can use the following CSV files available in plugin-data for testing:
`test/test-files/data-project/data/bulkUpsertLarge.csv` (big)
`test/test-files/data-project/data/bulkUpsert.csv` (smol)

### What issues does this PR fix or reference?
@W-13656292@

cristiand391 commented on src/bulkDataRequestCache.ts:
```diff
@@ -188,6 +188,28 @@ export class BulkUpsertRequestCache extends BulkDataRequestCache {
+      isState: true,
+      filename: BulkImportRequestCache.getFileName(),
+      stateFolder: Global.SF_STATE_FOLDER,
+      ttl: Duration.days(7),
```
bulk ingest/query job results are available for 7 days after being created:
https://developer.salesforce.com/docs/atlas.en-us.252.0.salesforce_app_limits_cheatsheet.meta/salesforce_app_limits_cheatsheet/salesforce_app_limits_platform_bulkapi.htm

cristiand391 commented on src/commands/data/import/bulk.ts:
```diff
@@ -0,0 +1,246 @@
+  jobId: string;
+  processedRecords?: number;
+  successfulRecords?: number;
+  failedRecords?: number;
```
JSON output:

**async**: jobID
**sync**: jobID and num of processed/successful/failed records

cristiand391 commented on src/commands/data/import/bulk.ts:
```diff
@@ -0,0 +1,246 @@
+    });
+
+    try {
+      await job.poll(5000, timeout.milliseconds);
```
5s for the polling interval, we could add a `--poll-interval` flag if users want to customize this later.

cristiand391 commented on src/commands/data/import/bulk.ts:
```diff
@@ -0,0 +1,246 @@
+        jobId: jobInfo.id,
+        processedRecords: jobInfo.numberRecordsProcessed,
+        successfulRecords: jobInfo.numberRecordsProcessed - (jobInfo.numberRecordsFailed ?? 0),
+        failedRecords: jobInfo.numberRecordsFailed,
```
The API only gives us the total of records processed (that includes success/failure) and failed ones so we need calculate successful qty.
https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/get_job_info.htm

cristiand391 commented on src/commands/data/import/bulk.ts:
```diff
@@ -0,0 +1,246 @@
+        ms.stop('failed');
+        throw messages.createError(
+          'error.jobFailed',
+          [jobInfo.errorMessage, conn.getUsername(), job.id],
```
`jobInfo.errorMessage` is guaranteed to be present on `state=Failed`:
https://github.com/jsforce/jsforce/blob/160426335c3d6f8efd1c3244eacb0454e755c988/src/api/bulk2.ts#L802
https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/get_job_info.htm

cristiand391 commented on src/commands/data/import/bulk.ts:
```diff
@@ -0,0 +1,246 @@
+      if (jobInfo.state === 'Aborted') {
+        ms.stop('failed');
+        // TODO: replace this msg to point to `sf data bulk results` when it's added (W-12408034)
+        throw messages.createError('error.jobAborted', [conn.getUsername(), job.id], [], err as Error);
```
aborted jobs don't include an error msg, we log the `sf org open ..` msg so they can view more details in the org.

cristiand391 commented on src/commands/data/import/resume.ts:
```diff
@@ -0,0 +1,193 @@
+      char: 'w',
+      unit: 'minutes',
+      summary: messages.getMessage('flags.wait.summary'),
+      defaultValue: 5,
```
`data import resume` has a default wait time of 5 minutes.
This is on purpose so that if you don't specify `--wait` you still have a good chance the job finishes on time.

`data upsert resume` and `data delete resume` have `--wait` with `0` as a default, so on a first run without `--wait` you have to re-run them again with a bigger timeout if the job is still running.

cristiand391 commented on src/commands/data/import/resume.ts:
```diff
@@ -0,0 +1,193 @@
+      char: 'i',
+      length: 18,
+      startsWith: '750',
+      exactlyOne: ['use-most-recent'],
```
following @VivekMChawla's comments about specificity:
https://salesforce-internal.slack.com/archives/G02K6C90RBJ/p1722277847033949

`sf data import resume` is invalid, you have to pass either `--job-id` or `--use-most-recent`

cristiand391 commented on src/commands/data/import/resume.ts:
```diff
@@ -0,0 +1,193 @@
+      flags['job-id'],
+      flags['use-most-recent'],
+      undefined,
+      undefined
```
the 4th arg is supposed to be `api-version` but I didn't add that flag here because the API version is cached by `data import bulk` so we don't need to pass it here.

cristiand391 commented on src/commands/data/import/resume.ts:
```diff
@@ -0,0 +1,193 @@
+            const numberRecordsFailed = data?.numberRecordsFailed ?? 0;
+
+            if (data?.numberRecordsProcessed) {
+              return (data.numberRecordsProcessed - numberRecordsFailed).toString();
```
we default to `0` on L84 for failed records if no API info is available yet (this state update happens while job is being processed) but only render the `Successful records` block  if the API returns `numberRecordsProcessed`.
This makes oclif/mso render a spinner instead of `0` when there's no processed data (first seconds of a job run or if it failed).

cristiand391 commented on src/commands/data/import/resume.ts:
```diff
@@ -0,0 +1,193 @@
+      const jobInfo = await job.check();
+
+      // send last data update so job status/num. of records processed/failed represent the last update
+      ms.goto('Processing the job', jobInfo);
```
if `job.poll` on L138 throws then the state rendered is from the previous poll update so we do one last update with fresh data from the org (`job.check()`) so the last state rendered is accurate (job state, record counter).

mdonnalley commented on src/commands/data/import/bulk.ts:
```diff
@@ -0,0 +1,246 @@
+    const baseUrl = flags['target-org'].getField<string>(Org.Fields.INSTANCE_URL).toString();
+
+    const ms = new MultiStageOutput<JobInfoV2>({
+      jsonEnabled: flags.json ?? false,
```
you should use `this.jsonEnabled()` since there's some env var logic in there: https://github.com/salesforcecli/sf-plugins-core/blob/main/src/sfCommand.ts#L174

mdonnalley commented on src/commands/data/import/bulk.ts:
```diff
@@ -0,0 +1,246 @@
+    });
+
+    if (async) {
+      ms.goto('Creating ingest job');
```
Do you need to use `.skipTo` or `.goto`?

`.goto` is going to mark any skipped stages as `complete` whereas `.skipTo` will mark any skipped stages as `skipped`

cristiand391 replied:
in async there's only one stage:
https://github.com/salesforcecli/plugin-data/blob/be3e6686163df2f823dfe240f8f9402c9e622ef1/src/commands/data/import/bulk.ts#L78

But I think you are right, we can leave all stages and skip them instead of not rendering them (`import resume` does this) 👍🏼

mdonnalley commented on src/commands/data/import/bulk.ts:
```diff
@@ -0,0 +1,246 @@
+      ms.goto('Processing the job', jobInfo);
+
+      if (jobInfo.numberRecordsFailed) {
+        ms.stop('failed');
```
just fyi, you could also use `ms.error()` here - it does the exact the same thing as `ms.stop('failed')`

https://github.com/oclif/multi-stage-output/blob/main/src/multi-stage-output.tsx#L316

mdonnalley commented on src/commands/data/import/resume.ts:
```diff
@@ -0,0 +1,193 @@
+    const conn = resumeOpts.options.connection;
+
+    const ms = new MultiStageOutput<JobInfoV2>({
+      jsonEnabled: flags.json ?? false,
```
`this.jsonEnabled()`

mdonnalley commented on src/commands/data/import/resume.ts:
```diff
@@ -0,0 +1,193 @@
+      jsonEnabled: flags.json ?? false,
+      stages: ['Creating ingest job', 'Processing the job'],
+      title: 'Importing data',
+      stageSpecificBlock: [
```
are all these blocks the same as the `bulk` one? If so, might be nice to put this into a single place, like what we did in PDR: https://github.com/salesforcecli/plugin-deploy-retrieve/blob/main/src/utils/deployStages.ts

mdonnalley commented:
**QA**

🔴  bulk import with --wait

```
❯ sf data bulk import -f ~/repos/salesforcecli/plugin-data/test/test-files/data-project/data/bulkUpsert.csv --sobject account --wait 10

 ›   Warning: @salesforce/plugin-data is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.

 ───────────────── Importing data ─────────────────

 ✔ Creating ingest job 4ms
 ✔ Processing the job 6.01s
   ▸ Processed records: 10
   ▸ Successful records: 10
   ▸ Failed records: ◝

 Status: JobComplete
 Job Id: 750Oy00000Dh2gcIAB
 Elapsed Time: 7.61s
```

It works as expected but `Failed records` is left with an unresolved spinner

🟡 bulk import

Question: is it supposed to be async by default?

```
❯ sf data bulk import -f ~/repos/salesforcecli/plugin-data/test/test-files/data-project/data/bulkUpsert.csv --sobject account
 ›   Warning: @salesforce/plugin-data is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.

 ───────────── Importing data (async) ─────────────

 ✔ Creating ingest job 4.05s

 Status: UploadComplete
 Job Id: 750Oy00000DgnMcIAJ
 Elapsed Time: 4.06s

Run "sf data import resume --job-id 750Oy00000DgnMcIAJ" to resume the operation.
```

🟢 bulk import with --async

```
❯ sf data bulk import -f ~/repos/salesforcecli/plugin-data/test/test-files/data-project/data/bulkUpsert.csv --sobject account --async
 ›   Warning: @salesforce/plugin-data is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.

 ───────────── Importing data (async) ─────────────

 ✔ Creating ingest job 1.41s

 Status: UploadComplete
 Job Id: 750Oy00000Dgu7yIAB
 Elapsed Time: 1.41s

Run "sf data import resume --job-id 750Oy00000Dgu7yIAB" to resume the operation.
```

🟢 resume an async import with --job-id

```
❯ sf data import resume --job-id 750Oy00000DgnMcIAJ
 ›   Warning: @salesforce/plugin-data is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.

 ───────────────── Importing data ─────────────────

 ◯ Creating ingest job - Skipped
 ✔ Processing the job 1.11s
   ▸ Processed records: 10
   ▸ Successful records: 10
   ▸ Failed records: 0

 Status: JobComplete
 Job Id: 750Oy00000DgnMcIAJ
 Elapsed Time: 1.13s
```

🟢 resume an async import with --use-most-recent

```
❯ sf data import resume --use-most-recent
 ›   Warning: @salesforce/plugin-data is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.

 ───────────────── Importing data ─────────────────

 ◯ Creating ingest job - Skipped
 ✔ Processing the job 1.11s
   ▸ Processed records: 10
   ▸ Successful records: 10
   ▸ Failed records: 0

 Status: JobComplete
 Job Id: 750Oy00000Dh2IPIAZ
 Elapsed Time: 1.14s
```

🟡 use invalid id for resume

I would have expected an error about the id not being valid but instead got this. Not sure if that's right or not

```
❯ sf data import resume --job-id 750Oy00000Dh2IPIAA
 ›   Warning: @salesforce/plugin-data is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.
Error (CannotCreateResumeOptionsWithoutAnOrgError): Cannot create a cache entry without a valid org.
```

🔴 links in terminal that doesn't support links

you need to provide your own fallback so that terminal-link doesn't insert non-visible whitespace characters, which is what I think causes this issue. See https://github.com/salesforcecli/plugin-deploy-retrieve/blob/main/src/utils/deployStages.ts#L78

![Oct-16-2024 12-53-54](https://github.com/user-attachments/assets/b3a5de9c-0278-4a09-8296-a4a81d6e0dd4)


🟢 async insert and resume with large csv

```
❯ sf data bulk import -f ~/repos/salesforcecli/plugin-data/test/test-files/data-project/data/bulkUpsertLarge.csv --sobject account --async
 ›   Warning: @salesforce/plugin-data is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.

 ───────────── Importing data (async) ─────────────

 ✔ Creating ingest job 3.29s

 Status: UploadComplete
 Job Id: 750Oy00000DgjAyIAJ
 Elapsed Time: 3.32s

Run "sf data import resume --job-id 750Oy00000DgjAyIAJ" to resume the operation.

~/repos/trailheadapps/dreamhouse-lwc on  main via ⬢ v20.15.0 took 5.7s
❯ sf data import resume --job-id 750Oy00000DgjAyIAJ
 ›   Warning: @salesforce/plugin-data is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.

 ───────────────── Importing data ─────────────────

 ◯ Creating ingest job - Skipped
 ✔ Processing the job 2m 30.23s
   ▸ Processed records: 76380
   ▸ Successful records: 76380
   ▸ Failed records: 0

 Status: JobComplete
 Job Id: 750Oy00000DgjAyIAJ
 Elapsed Time: 2m 30.79s

```

🟢 handles failures

```
❯ sf data bulk import -f ~/repos/salesforcecli/plugin-data/test/test-files/data-project/data/bulkUpsert.csv --sobject account --line-ending CRLF --wait 10
 ›   Warning: @salesforce/plugin-data is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.

 ───────────────── Importing data ─────────────────

 ✔ Creating ingest job 3ms
 ✘ Processing the job 11.83s
   ▸ Processed records: ✘
   ▸ Successful records: ✘
   ▸ Failed records: ✘

 Status: Failed
 Job Id: 750Oy00000Dh2gdIAB
 Elapsed Time: 13.42s

Error (JobFailedError): Job failed to be processed due to:
ClientInputError : LineEnding is invalid on user data. Current LineEnding setting is CRLF

To review the details of this job, run:
sf org open --target-org test-f7zxeasp0pzx@example.com --path "/lightning/setup/AsyncApiJobStatus/page?address=%2F750Oy00000Dh2gdIAB"
```

cristiand391 commented:
> It works as expected but Failed records is left with an unresolved spinner

should be fixed now (`data import resume` was handling it correctly, all MSO stuff is merged into one class now)

> 🟡 bulk import
Question: is it supposed to be async by default?

yes, same as `data bulk delete/upsert`. I can't find why that decision was made.
`data import resume` does have a default wait time on purpose:
https://github.com/salesforcecli/plugin-data/pull/1091#discussion_r1803126186

> 🟡 use invalid id for resume

fixed by writing a separate cache resolver for `data import` like we did for `data export`.

> 🔴 links in terminal that doesn't support links

added the fallback

cristiand391 commented:
@mdonnalley this is ready for review/qa again, thanks!

also, please squash-merge the PR when it's good to go (bunch of commits in my branch)

mdonnalley commented:
🟢 bulk import with --wait
```
❯ sf data bulk import -f ~/repos/salesforcecli/plugin-data/test/test-files/data-project/data/bulkUpsert.csv --sobject account --wait 10

 ›   Warning: @salesforce/plugin-data is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.

 ───────────────── Importing data ─────────────────

 ✔ Creating ingest job 1.49s
 ✔ Processing the job 11.32s
   ▸ Processed records: 10
   ▸ Successful records: 10
   ▸ Failed records: 0

 Status: JobComplete
 Job Id: 750Ei00000EQhxKIAT
 Elapsed Time: 12.93s
```

🟢 use invalid id for resume
```
❯ sf data import resume --job-id 750Ei00000EQjmGIAz
 ›   Warning: @salesforce/plugin-data is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.
Error (BulkRequestIdNotFoundError): Could not find a cache entry for job ID 750Ei00000EQjmGIAz.
```

🟢 links in terminal that doesn't support links
